### ISSUE TITLE
Question about IPI (Inter-Processor Interrupt) in SMP Linux+OP-TEE

### ISSUE THREAD

**Issue by gagachang** [OP-TEE/optee_os#6767]
Hello,

I have some questions about IPI in SMP Linux+OP-TEE.

To be specific, if there are four CPUs run Linux and OP-TEE at EL1, in SMP manner.
User runs an CA from Linux CLI on CPU-0, causing CPU-0 enters OP-TEE to run TA.
Suddenly, CPU-1, which runs Linux, want to send IPI to "all the CPUs".

1.  Will CPU-1 also send IPI to CPU-0, which is currently in secure world?
2. If question one is true, how CPU-1 handles that IPI ? Will it issue RPC to go back to Linux ?

I am wondering if there are some slides or documentations which describe the IPI handling in normal world and secure world,
Thanks!

**Comment by jenswi-linaro** [comment]
> * Will CPU-1 also send IPI to CPU-0, which is currently in secure world?

Yes, it will be received in the secure world as a "foreign" interrupt when foreign interrupts are unmasked, which they are while executing in a thread context or in a TA.

> * If question one is true, how CPU-1 handles that non-secure IPI ? Will it issue RPC to go back to Linux ?

Yes, the state will be saved in a thread context. An RPC will report the thread ID so it can be resumed later. The thread can be resumed on a different CPU so migration of a thread between CPUs is possible.

You can find some details here.
https://optee.readthedocs.io/en/latest/architecture/core.html#interrupt-handling

**Comment by jenswi-linaro** [comment]
Non-secure IPI is not different from other interrupts from OP-TEE point of view.

**Comment by gagachang** [comment]
Thank you @jenswi-linaro ! It's so clear!
I will read that section.
